### ISSUE TITLE
patHint should warn of redundant bang patterns for tuple and list constructors

### ISSUE THREAD

**Issue by tomjaguarpaw** [ndmitchell/hlint#257]
`patHint` should warn of redundant bang patterns for `PTuple` and `PList` (and probably `PRec` too) and recurse into `PAsPat` (and probably into `PatTypeSig`).

https://github.com/ndmitchell/hlint/blob/a6be5c52f4025d7269d87e860bd4e18bec126712/src/Hint/Structure.hs#L109

**Comment by ndmitchell** [comment]
Can you give an example of something it doesn't hint at for tuple/list, but which it should? Agree on the other two pieces.

**Comment by tomjaguarpaw** [comment]
All of these should have warnings

````haskell
data Tuple a b = Tuple a b

data List a = Nil | Cons a (List a)

f !(a, b) = ()

g ![a] = ()

h !(Tuple a b) = ()

i !(Cons a Nil) = ()
````

but I only get

````
tmp.hs:9:3: Warning: Redundant bang pattern
Found:
  !(Tuple a b)
Why not:
  (Tuple a b)

tmp.hs:11:3: Warning: Redundant bang pattern
Found:
  !(Cons a Nil)
Why not:
  (Cons a Nil)

2 hints
````

**Comment by ndmitchell** [comment]
Agreed. Pull requests most welcome :)

**Comment by tomjaguarpaw** [comment]
Great.  I plan to work on it.

**Comment by ndmitchell** [comment]
In the original comment you mentioned `PAsPat` and `PatTypeSig`, but the patch didn't cover them. Was that deliberate? Is there more work to do on this ticket?

**Comment by tomjaguarpaw** [comment]
I tested `PatTypeSig` and discovered `hlint` already coped with it fine.  I assumed that was a consequence of how uniplate works descending into nested patterns.

I overlooked `PAsPat` though.  I'll check it now.

**Comment by tomjaguarpaw** [comment]
I got a bit confused.  I am testing the following file

````haskell
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE BangPatterns #-}

data Tuple a b = Tuple a b

data List a = Nil | Cons a (List a)

data Rec a = Rec { field :: a }

e !() = ()

f !(a, b) = ()

g ![a] = ()

h !(Tuple a b) = ()

i !(Cons a Nil) = ()

j !Rec { field = a } = ()

k !x@() = ()

l (!(() :: ())) = ()
````

I get the result

````
/tmp/bangs.hs:10:3: Warning: Redundant bang pattern
Found:
  !()
Why not:
  ()

/tmp/bangs.hs:12:3: Warning: Redundant bang pattern
Found:
  !(a, b)
Why not:
  (a, b)

/tmp/bangs.hs:14:3: Warning: Redundant bang pattern
Found:
  ![a]
Why not:
  [a]

/tmp/bangs.hs:16:3: Warning: Redundant bang pattern
Found:
  !(Tuple a b)
Why not:
  (Tuple a b)

/tmp/bangs.hs:18:3: Warning: Redundant bang pattern
Found:
  !(Cons a Nil)
Why not:
  (Cons a Nil)

/tmp/bangs.hs:20:3: Warning: Redundant bang pattern
Found:
  !Rec{field = a}
Why not:
  Rec{field = a}

/tmp/bangs.hs:22:3: Warning: Redundant bang pattern
Found:
  !x@()
Why not:
  x@()

7 hints
````

So it does work for `PAsPat` (because that's *already* in the code!)  but not `PatTypeSig` (so nothing to do with uniplate).  I'll add the latter when I get the chance.

**Comment by tomjaguarpaw** [comment]
Are there any tests for this stuff, by the way?  I couldn't find any.

**Comment by ndmitchell** [comment]
See https://github.com/ndmitchell/hlint/blob/master/src/Hint/Pattern.hs#L27-L45 - those get run by the test code. If you load up in `ghci`, then `:test` will extract/run them, as well the various CI things.

**Comment by tomjaguarpaw** [comment]
Can you explain how those tests work?  I don't understand it.

Also, how should I load it in GHCi?  This is what I get when I try:

````
% ghci  -isrc src/Hint/Pattern.hs
GHCi, version 7.6.3: http://www.haskell.org/ghc/  :? for help
Loading package ghc-prim ... linking ... done.
Loading package integer-gmp ... linking ... done.
Loading package base ... linking ... done.

<no location info>: can't find file: src\Main.hs
Failed, modules loaded: none.

src/CmdLine.hs:23:8:
    Could not find module `Paths_hlint'
    Use -v to see a list of the files searched for.
Failed, modules loaded: none.
````

**Comment by ndmitchell** [comment]
The tests work by extracting the sections inside `<test>`, see https://github.com/ndmitchell/hlint/blob/master/src/Test/Annotations.hs for the code that actually does the needful.

To load in ghci, just type `ghci`. There is a `.ghci` file that does the necessary includes and Paths files.

**Comment by tomjaguarpaw** [comment]
Is there a strange Windowsism going on here?  I had to change back slashes to forward slashes: https://github.com/ndmitchell/hlint/blob/master/.ghci#L4

**Comment by ndmitchell** [comment]
Probably - I've never tested that on Linux. They probably should be forward slashes. I think `cabal test` will do it too, if run from the root of the repo.

**Comment by tomjaguarpaw** [comment]
OK, I see how that works.  The bit in the comments is the suggestion we are expecting.  OK, I will add some tests when I do the `PatTypeSig`.

**Comment by ndmitchell** [comment]
I've raised #260 to actually document all this stuff.
